pushapkscript/task.py

```diff
@@ -8,15 +8,30 @@
 
 
 def extract_android_product_from_scopes(context):
-    prefix = context.config['taskcluster_scope_prefix']
+    prefixes = _get_scope_prefixes(context)
+    all_scopes_starting_with_all_prefixes = [
+         scope
+         for scope in context.task['scopes']
+         for prefix in prefixes
+         if scope.startswith(prefix)
+    ]
+
     scope = get_single_item_from_sequence(
-        context.task['scopes'],
-        condition=lambda scope: scope.startswith(prefix),
+        all_scopes_starting_with_all_prefixes,
+        condition=lambda _: True,     # it must just contain 1 single item
```

Comment:
IMHO, as a reader, the important part here is that there should be a single item - unfortunately, this constraint is encoded within a very large call to `get_single_item_from_sequence` (it seems like the important part is the `condition=lambda _: True`, which also isn't super clear to the next reader of the code).
I understand that we follow a similar pattern of `get_single_item_from_sequence` elsewhere, but perhaps this is more cleanly solved with an assert? `assert len(all_scopes_starting_with_all_prefixes) == 1` :thinking:

Reply:
Yeah, I agree. The condition makes this call odd. The reason why `get_single_item_from_sequence()` was created is because too many *scripts were [reimplementing the same logic](https://github.com/mozilla-releng/scriptworker/pull/195). For instance, `get_single_item_from_sequence()` [errors out whenever](https://github.com/mozilla-releng/scriptworker/blob/931a23b64e40cd913a629b65f6e9ade989a33765/scriptworker/utils.py#L689-L700) `len(all_scopes_starting_with_all_prefixes) != 1`. I'm not a huge fan of putting this logic back in pushapkscript. How about: 

```py
def extract_android_product_from_scopes(context):
    prefixes = _get_scope_prefixes(context)

    scope, _ = get_single_item_from_sequence(
        sequence=[(scope, prefix) for scope in context.task['scopes'] for prefix in prefixes],
        condition=lambda scope_then_prefix: scope_then_prefix[0].startswith(scope_then_prefix[1]),
        ErrorClass=TaskVerificationError,
        no_item_error_message='No scope starting with any of these prefixes {} found'.format(prefixes),
        too_many_item_error_message='More than one scope matching these prefixes {} found'.format(prefixes),
    )

    android_product = scope.split(':')[-1]

    return android_product

```

? Does this look still readable/not too hacky to you? 

Side note regarding `assert` statements: they usually get [optimized away in production](https://stackoverflow.com/questions/944592/best-practice-for-python-assert) ([docs](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement)). The way to error our would be to raise exceptions.

Reply:
Mmmm, I wonder if I shouldn't change https://github.com/mozilla-releng/scriptworker/blob/931a23b64e40cd913a629b65f6e9ade989a33765/scriptworker/utils.py#L689 to expand sequences within the sequence.

For instance:

```py
def extract_android_product_from_scopes(context):
    # ...
    scope, _ = get_single_item_from_sequence(
        sequence=[(scope, prefix) for scope in context.task['scopes'] for prefix in prefixes],
        condition=lambda scope, prefix: scope.startswith(prefix),
        # ...
    )
    # ...
```

Reply:
`get_single_item_from_sequence` is bugging me in a weird way - I feel like it's not very pythonic. It seems like it's generalizing some straightforward code that has the potential to have a _lot_ of options - already, it's getting parameters like `append_sequence_to_error_message`, which is a bit of a code smell, IMHO.

If we start adding features like automatically expanding nested sequences (and putting an option behind that, too), then we have some fallout:
* Complexity of `get_single_item_from_sequence` increases dramatically
* As a consumer of `get_single_item_from_sequence`, if I'm confused by the result and I investigate the implementation, I see a _lot_ of code (and, since there's a lot of boolean options, only ~half the code is relevant for my current use case)

My theory is that `get_single_item_from_sequence` was created to enhance reusability, but I feel like it's just increasing complexity and lowering readability. Besides, since it's so general, almost all of its behaviour needs to be passed in anyways (filtering logic, the exact error message for each case, etc). Is it worth it, and (:apple: for :thought_balloon:) what would we lose if we inlined the code in each case? So, here, it would be:

```
def extract_android_product_from_scopes(context):
    prefixes = _get_scope_prefixes(context)
    scopes = context.task['scopes']
    filtered_scopes = [scope for scope in scopes for prefix in prefixes if scope.startswith(prefix)]

    if len(filtered_scopes) == 0:
        raise TaskVerificationError('No scope starting with any of these prefixes: {}. Given scopes: {}'
                                    .format(prefixes, scopes))
    if len(filtered_scopes) > 1:
        raise TaskVerificationError('More than one scope found. Given scopes: {}'
                                    .format(scopes))

    android_product = filtered_scopes[0].split(':')[-1]

    return android_product
```

Note the benefits of de-generalizing this:
* The "empty" error message is improved. At the end of the line, where we dump the "given" section, we don't show all the `(scope, prefix)` combinations, but rather just the list of scopes. This is good, since we're already dumping the list of prefixes earlier in the error message
* Complexity is lowered, we don't need to mentally omit all the `if` that don't apply to us - we can just remove them from the code.
* Readability is enhanced, since all the behaviour is presented to us, and none is hidden in a helper function.

-----

Sorry for the essay! Feel free to merge this PR whenever, it's definitely :ok_hand:. This conversation is almost entirely a ":apple: for :thought_balloon:"-er

Reply:
Your arguments make sense. I don't have a lot of counter-arguments left :sweat_smile:. Let's ask Aki, what he thinks of this. I opened https://github.com/mozilla-releng/pushapkscript/issues/60 to do so. In the meantime, I'm merging the PR as is.